### PR TITLE
dwarf: make debug_line header parser more resilient

### LINKED ISSUE #2449: Delve Crashing
Please answer the following before submitting your issue:

Note: Please include any substantial examples (debug session output,
stacktraces, etc) as linked gists.

1. What version of Delve are you using (`dlv version`)? 1.6.0
2. What version of Go are you using? (`go version`)? 1.16.2
3. What operating system and processor architecture are you using? Windows, amd64
4. What did you do? Trying to debug my application and the delve immediately crashes
5. What did you expect to see? No crash
6. What did you see instead? Crash

Here is the crash:

```
> C:\Users\User\go\bin\dlv.exe debug .
panic: runtime error: index out of range [1] with length 0

goroutine 12 [running]:
encoding/binary.littleEndian.Uint16(...)
        C:/tmprun/apr142021/test2/myproj/.deps/go/src/encoding/binary/binary.go:53
github.com/go-delve/delve/pkg/dwarf/line.(*formReader).next(0xc000ca07e0, 0xc0008e8060, 0x0)
        C:/Users/User/go/pkg/mod/github.com/go-delve/delve@v1.6.0/pkg/dwarf/line/parse_util.go:93 +0x811
github.com/go-delve/delve/pkg/dwarf/line.parseFileEntries5(0xc000ca06c0, 0xc0008e8060)
        C:/Users/User/go/pkg/mod/github.com/go-delve/delve@v1.6.0/pkg/dwarf/line/line_parser.go:234 +0x105
github.com/go-delve/delve/pkg/dwarf/line.Parse(0xc0005cd530, 0x2b, 0xc0008e8060, 0x0, 0x0, 0x1, 0x8, 0xc000ca0510)
        C:/Users/User/go/pkg/mod/github.com/go-delve/delve@v1.6.0/pkg/dwarf/line/line_parser.go:93 +0x32c
github.com/go-delve/delve/pkg/proc.(*BinaryInfo).loadDebugInfoMaps(0xc000060140, 0xc0006e20c0, 0xc000ed0000, 0x5136d8, 0x513800, 0xc0013e4000, 0x1fa151, 0x1fa200, 0xc0006cf750, 0x0)
        C:/Users/User/go/pkg/mod/github.com/go-delve/delve@v1.6.0/pkg/proc/bininfo.go:1610 +0x5d8
created by github.com/go-delve/delve/pkg/proc.loadBinaryInfoPE
        C:/Users/User/go/pkg/mod/github.com/go-delve/delve@v1.6.0/pkg/proc/bininfo.go:1315 +0x555

```

### PR DESCRIPTION
**Pull request by aarzilli**
Check for errors, log them and return early, do not try to allocate
large chunks of memory that we can never possibly read from the file.

Fixes #2449

### CODE REVIEW

File: pkg/dwarf/line/line_parser.go [hunk 1]
```diff
@@ -168,26 +183,44 @@ func parseIncludeDirs5(info *DebugLineInfo, buf *bytes.Buffer) {
...
 	entry := new(FileEntry)
 
-	entry.Path, _ = util.ParseString(buf)
+	var err error
+	entry.Path, err = util.ParseString(buf)
+	if err != nil {
+		if info.Logf != nil {
+			info.Logf("error reading file entry: %v", err)
+		}
+		return nil
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/dwarf/line/line_parser.go:225]: Nope. Changed.

File: pkg/dwarf/line/line_parser.go [hunk 2]
```diff
@@ -135,19 +143,26 @@ func parseDebugLinePrologue(dbl *DebugLineInfo, buf *bytes.Buffer) {
...
 
 		info.IncludeDirs = append(info.IncludeDirs, str)
 	}
+	return true
 }
 
 // parseIncludeDirs5 parses the directory table for DWARF version 5.
-func parseIncludeDirs5(info *DebugLineInfo, buf *bytes.Buffer) {
+func parseIncludeDirs5(info *DebugLineInfo, buf *bytes.Buffer) bool {
 	dirEntryFormReader := readEntryFormat(buf, info.Logf)
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/dwarf/line/line_parser.go:166]: Done.

**Comment by aarzilli** [comment]
@nd it looks like the TeamCity builders are stuck?

**Comment by nd** [comment]
found a problem, hopefully will be fixed soon

**Comment by aarzilli** [comment]
When something like this happens how do you want me to report it?

**Comment by nd** [comment]
Builds are finally running again, sorry to took so long. TC in cloud is publicly launched today and there were some unexpected problems. 

In case it happen again you can file an issue [here](https://youtrack.jetbrains.com/newIssue?project=TCC).

**Comment by aarzilli** [comment]
> Builds are finally running again, sorry to took so long. TC in cloud is publicly launched today and there were some unexpected problems.
> 
> In case it happen again you can file an issue [here](https://youtrack.jetbrains.com/newIssue?project=TCC).

Thanks, will do.

**Review by hyangah — APPROVED** [review]

**Review by derekparker — APPROVED** [review]
LGTM

**Pull request merged**